safedelete/managers.py

```diff
@@ -24,6 +23,90 @@ def undelete(self):
+        models. It is not recommended to use `force_visibility` outside of related
+        models because it will create a new queryset.
+
+        Args:
+            force_visibility: Show deleted models (default: {False})
+        """
+        if force_visibility is not None:
+            self._safedelete_force_visibility = force_visibility
+        return super(SafeDeleteQueryset, self).all()
+
+    def check_field_filter(self, **kwargs):
+        """Check if the visibility for DELETED_VISIBLE_BY_FIELD needs t be put into effect.
+
+        DELETED_VISIBLE_BY_FIELD is a temporary visibility flag that changes
+        to DELETED_VISIBLE once asked for the named parameter defined in
+        `_safedelete_force_visibility`. When evaluating the queryset, it will
+        then filter on all models.
+        """
+        if self._safedelete_visibility == DELETED_VISIBLE_BY_FIELD \
+                and self._safedelete_visibility_field in kwargs:
+            self._safedelete_force_visibility = DELETED_VISIBLE
+
+    def filter(self, *args, **kwargs):
+        self.check_field_filter(**kwargs)
+        return super(SafeDeleteQueryset, self).filter(*args, **kwargs)
+
+    def get(self, *args, **kwargs):
+        self.check_field_filter(**kwargs)
+        return super(SafeDeleteQueryset, self).get(*args, **kwargs)
+
+    def filter_visibility(self):
+        """Add deleted filters to the current QuerySet.
+
+        Unlike QuerySet.filter, this does not return a clone.
+        This is because QuerySet._fetch_all cannot work with a clone.
+        """
+        force_visibility = getattr(self, '_safedelete_force_visibility', None)
+        visibility = force_visibility \
+            if force_visibility is not None \
+            else self._safedelete_visibility
+        if visibility in (DELETED_INVISIBLE, DELETED_VISIBLE_BY_FIELD, DELETED_ONLY_VISIBLE):
+            assert self.query.can_filter(), \
+                "Cannot filter a query once a slice has been taken."
+
+            # Add a query manually, QuerySet.filter returns a clone.
+            # QuerySet._fetch_all cannot work with clones.
+            self.query.add_q(
+                Q(
+                    deleted__isnull=visibility in (
+                        DELETED_INVISIBLE, DELETED_VISIBLE_BY_FIELD
+                    )
+                )
+            )
+
+    def __getattribute__(self, name):
+        """Methods that do not return a QuerySet should call filter_visibility first."""
+        attr = object.__getattribute__(self, name)
+        # These methods evaluate the queryset and therefore need to filter the
+        # visiblity set.
+        evaluation_methods = (
+            '_fetch_all', 'count', 'exists', 'aggregate', 'update', '_update',
+            'delete', 'undelete',
+        )
+        if hasattr(attr, '__call__') and name in evaluation_methods:
+            def decorator(*args, **kwargs):
+                self.filter_visibility()
```

Comment:
This needs to be documented in the docs for when using custom QuerySets. When working with the QuerySet's `query` directly that hasn't had the deleted filter applied yet, `filter_visibility` needs to be called first.